File: 프로그래머스/lv2/42842. 카펫/카펫.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> solution(int brown, int yellow) {
    vector<int> answer;
    std::vector<std::vector<int>> v;
    
    for (int i = 1; i <= yellow; i++) {
        if (yellow%i == 0) {
            std::vector<int> temp;
            temp.emplace_back(i);
            temp.emplace_back(yellow/i);
            v.emplace_back(temp);
        }
    }
    for (int i = 0; i < v.size(); i++) {
        if (v[i][0]*2 + v[i][1]*2 + 4 == brown) {
            answer.emplace_back(v[i][1]+2);
            answer.emplace_back(v[i][0]+2);
            break;
        }
    }
    return answer;
}

/*
24 =    1*24 -> 1*2 + 24*2 + 4
        2*12 -> 2*2 + 12*2 + 4
        3*8  -> 3*2 + 8*2 + 4
        4*6  -> 4*2 + 6*2 + 4
*/

File: 프로그래머스/2/42842. 카펫/카펫.java
class Solution {
    public int[] solution(int brown, int yellow) {
        int[] answer = new int[2];
        int r = 0;
        int c = 0;
        // r*2 + c*2 - 4 = brown
        // (r-2) * (c-2) = yellow
        
        for (int i = 3; i < 3000; i++) {
            for (int j = i; j < 2000000; j++) {
                if (j*2 + i*2 == brown + 4 && j*i - 2*j - 2*i + 4 == yellow) {
                    r = j;
                    c = i;
                    answer[0] = r;
                    answer[1] = c;
                    return answer;
                }
            }
        }
        
        answer[0] = r;
        answer[1] = c;
        return answer;
    }
}